### PR TITLE
Http stream end to end tests

### PR DESCRIPTION
**Pull request by ejizba**
These tests include:
- A "hello world" request with "transfer-encoding" set to "chunked" (the default for any streamed request)
- Streaming random data up to 2g in both the request and the response
- 256 concurrent requests

These tests are not meant to be exhaustive performance/stress testing and do not represent the limits of what the feature can do, just the limits of what I wanted to included in our end-to-end tests. I still want separate perf testing

I'm still testing streaming both on and off, with the former as the default and the latter as a part of the "oldConfig" test pass.

### CODE REVIEW

File: azure-pipelines/test.yml [hunk 1]
```diff
@@ -89,9 +89,10 @@ jobs:
   - script: mv *.tgz package.tgz
     displayName: 'Rename v4 tgz file' # because the publish command below requires an exact path
     workingDirectory: '$(Pipeline.Workspace)/nodeLibraryV4/drop'
-  - script: npm install "$(Pipeline.Workspace)/nodeLibraryV4/drop/package.tgz"
-    displayName: 'install latest v4 library'
-    workingDirectory: 'app/v4'
+  # TODO: Temporary change until actual v4 drop exists with http stream changes
```
Comment (ejizba) [inline azure-pipelines/test.yml:92]: Will revert these changes and remove "app/v4/azure-functions-4.1.0.tgz" after stream support is merged in the library package and before merging this PR

File: app/v3/utils/getRandomTestData.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+export function getRandomInt(min: number, max: number): number {
+    return Math.floor(Math.random() * (max - min) + min);
+}
+
+export function getRandomBoolean(percentTrue: number): boolean {
+    return Math.random() * 100 > percentTrue;
+}
+
+export async function addRandomDelay(): Promise<void> {
+    if (getRandomBoolean(95)) {
```
Comment (castrodd) [inline app/v3/utils/getRandomTestData.ts:27]: I have a couple questions here. First, why set the parameter to getRandomBoolean() to 95? Second, why add the if/else blocks and not just go straight to delay(getRandomInt(0, 250))?

File: app/v4/src/utils/getRandomTestData.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License.
```
Comment (castrodd) [inline app/v4/src/utils/getRandomTestData.ts:2]: Why do we need two copies of this file for v3 and v4? Would it be possible to create a util folder at app/ that they can both use?

File: src/utils/getRandomTestData.ts [hunk 4]
```diff
@@ -1,7 +1,8 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT License.
```
Comment (castrodd) [inline src/utils/getRandomTestData.ts:2]: I might be missing the reason why we need to do this, but if possible it would be nice to avoid duplicating this code in multiple places.

File: src/global.test.ts [hunk 5]
```diff
@@ -123,7 +130,9 @@ async function startFuncProcess(appPath: string): Promise<void> {
 
     childProc.stdout?.on('data', (data: string | Buffer) => {
         data = data.toString();
-        process.stdout.write(data);
+        if (!funcCliSettings.hideOutput) {
```
Comment (castrodd) [inline src/global.test.ts:133]: Would it be clearer to name this setting `showOutput`?

Comment (ejizba) [reply src/global.test.ts:133]: My general philosophy is to make the default value false (to match `undefined` which is false-y) and then choose the name based on that. So I couldn't do `showOutput` since then the default would have to be true

File: app/v3/utils/getRandomTestData.ts (hunk 2 above)
Comment (ejizba) [reply app/v3/utils/getRandomTestData.ts:27]: I'm trying to simulate somewhat realistic scenarios and I wanted a combination of asynchronous and synchronous delays. 95 means return true 95% of the time, which means the vast majority of the time it will be an async delay. I hope most of our users use async, plus if I do sync too often the test will timeout (running tests in parallel with async is inherently much faster)

File: app/v4/src/utils/getRandomTestData.ts (hunk 3 above)
Comment (ejizba) [reply app/v4/src/utils/getRandomTestData.ts:2]: Well they're completely separate projects that just happen to live in the same git repo. By default you can't reference a file outside the project's root folder. Here's an example error if I try to reference a file outside the project root:

> Error: File '/Users/erijiz/Repos/js-e2e-tests/src/utils/getRandomTestData.ts' is not under 'rootDir' '/Users/erijiz/Repos/js-e2e-tests/app/v4'. 'rootDir' is expected to contain all source files.

There's probably some magic settings we could use in the tsconfig.json to make this work, but I don't know how to do it. It's rare for me to have multiple projects in one git repo, which is why I've never been bothered enough to figure it out

File: app/v3/utils/getRandomTestData.ts (hunk 2 above)
Comment (castrodd) [reply app/v3/utils/getRandomTestData.ts:27]: Ah, so we're saying it's not just random in terms of the length of time, but also in the nature of the delay. Would you be open to a name change suggestion? Maybe, `addRandomAsyncOrSyncDelay`? That's very C#-ish, I know.

File: src/global.test.ts (hunk 5 above)
Comment (castrodd) [reply src/global.test.ts:133]: I see. Okay, no problem. I'll try to match that approach.

File: app/v3/utils/getRandomTestData.ts (hunk 2 above)
Comment (ejizba) [reply app/v3/utils/getRandomTestData.ts:27]: fixed

File: .vscode/settings.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
     "azureFunctions.showProjectWarning": false,
-    "editor.formatOnSave": true,
+    "editor.formatOnSave": false,
```
Comment (castrodd) [inline .vscode/settings.json:3]: I thought I commented on this earlier, but now I don't see it in the conversation history. Are we turning this off because it conflicts with another action (possibly the next line?)?

File: app/v4/src/utils/streamHttp.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            nextLogTime = Date.now() + logInterval;
+            console.log(`Progress: ${convertBToMb(bytesReceived)}mb`);
+        }
+
+        bytesReceived += chunk.length;
+    }
+    return bytesReceived;
+}
+
+const oneMb = 1024 * 1024;
```
Comment (castrodd) [inline app/v4/src/utils/streamHttp.ts:49]: Nit: can we move this constant to the top of the file? It is used before it is defined.

File: src/http.test.ts [hunk 8]
```diff
@@ -57,9 +65,105 @@ describe('http', () => {
         expect(body).to.equal('');
         expect(response.status).to.equal(200);
         const cookies = response.headers.get('Set-Cookie');
-        expect(cookies).to.equal(
-            'mycookie=myvalue; max-age=200000; path=/, mycookie2=myvalue; max-age=200000; path=/, mycookie3-expires=myvalue3-expires; max-age=0; path=/, mycookie4-samesite-lax=myvalue; path=/; samesite=lax, mycookie5-samesite-strict=myvalue; path=/; samesite=strict'
-        );
+        if (isOldConfig || model === 'v3') {
+            expect(cookies).to.equal(
+                'mycookie=myvalue; max-age=200000; path=/, mycookie2=myvalue; max-age=200000; path=/, mycookie3-expires=myvalue3-expires; max-age=0; path=/, mycookie4-samesite-lax=myvalue; path=/; samesite=lax, mycookie5-samesite-strict=myvalue; path=/; samesite=strict'
```
Comment (castrodd) [inline src/http.test.ts:70]: Is there a way to make clearer the difference between these two strings? It is hard at first glance to see how they differ.

**Review by castrodd — APPROVED** [review]

File: .vscode/settings.json (hunk 6 above)
Comment (ejizba) [reply .vscode/settings.json:3]: Yeah eslint (the action from line 4) wants this space gone, but "formatOnSave" was adding it back.

Screenshot 2024-02-05 at 2 29 09 PM

I can follow up separately from this PR to figure out ideal settings - I think the problem with "codeActionsOnSave" is that it only formats TypeScript files and it would be nice to format others files like json automatically as well. But suffice to say the conflict was super annoying lol
